src/controlplane/client/connector.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	if mTLSConfig.TLSSecretName == "" {
+		return nil, fmt.Errorf("mTLS secret name cannot be empty")
+	}
+
+	namespace := mTLSConfig.TLSSecretNamespace
+	secretName := mTLSConfig.TLSSecretName
+	if namespace == "" {
+		dp.logger.Debugf("TLS Secret name configured, but not TLS Secret namespace. Defaulting to `default` namespace.")
+		namespace = "default"
+	}
+
+	secret, err := dp.kc.CoreV1().Secrets(namespace).Get(context.Background(), secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not find secret %q containing TLS configuration: %w", secretName, err)
+	}
+
+	var cert, key, cacert []byte
+
+	var ok bool
+	if cert, ok = secret.Data["cert"]; !ok {
+		return nil, fmt.Errorf("could not find TLS certificate in `cert` field in secret %q", secretName)
+	}
+
+	if key, ok = secret.Data["key"]; !ok {
+		return nil, fmt.Errorf("could not find TLS key in `key` field in secret %q", secretName)
+	}
+
+	cacert, hasCACert := secret.Data["cacert"]
+	insecureSkipVerifyRaw, hasInsecureSkipVerify := secret.Data["insecureSkipVerify"]
+
+	if !hasCACert && !hasInsecureSkipVerify {
+		return nil, fmt.Errorf("both cacert and insecureSkipVerify are not set. One of them need to be set to be able to call ETCD metrics")
+	}
+
+	// insecureSkipVerify is set to false by default, and can be overridden with the insecureSkipVerify field
+	insecureSkipVerify := false
+	if hasInsecureSkipVerify {
+		insecureSkipVerify = strings.ToLower(string(insecureSkipVerifyRaw)) == "true"
+	}
+
+	tlsConfig, err := parseTLSConfig(cert, key, cacert, insecureSkipVerify)
+	if err != nil {
+		return nil, fmt.Errorf("parsing TLS config: %w", err)
+	}
+
+	return tlsConfig, nil
+}
+
+func parseTLSConfig(certPEMBlock, keyPEMBlock, cacertPEMBlock []byte, insecureSkipVerify bool) (*tls.Config, error) {
+	cert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	clientCertPool := x509.NewCertPool()
+
+	if len(cacertPEMBlock) > 0 {
+		clientCertPool.AppendCertsFromPEM(cacertPEMBlock)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{cert},
+		RootCAs:            clientCertPool,
+		InsecureSkipVerify: insecureSkipVerify,
+	}
+
+	tlsConfig.BuildNameToCertificate()
+
+	return tlsConfig, nil
+}
+
+func validateEndpointConfig(endpoints []config.Endpoint) error {
+	for _, e := range endpoints {
+		if _, err := url.Parse(e.URL); err != nil {
+			return fmt.Errorf("parsing endpoint url %q: %w", e.URL, err)
+		}
+
+		if err := validateAuth(e.Auth); err != nil {
+			return fmt.Errorf("validating auth for endpoint url %q: %w", e.URL, err)
+		}
+	}
+
+	return nil
+}
+
+func validateAuth(auth *config.Auth) error {
+	if auth == nil {
+		return nil
+	}
+
+	switch {
+	case strings.EqualFold(auth.Type, bearerAuth):
+		break
+	case strings.EqualFold(auth.Type, mTLSAuth):
+		return validateMTLS(auth.MTLS)
+	default:
+		return fmt.Errorf("authorization type not supported: %q", auth.Type)
```

Comment:
We are doing this twice, right? I'd get rid of early validation and just fail when we attempt to use it.